x/gamm/keeper/migrate.go

```diff
@@ -109,27 +109,46 @@ func (k Keeper) GetLinkedBalancerPoolID(ctx sdk.Context, concentratedPoolId uint
 	return sdk.BigEndianToUint64(balancerPoolIdBigEndian), nil
 }
 
-// OverwriteMigrationRecords sets the balancer to gamm pool migration info to the store
-// Deletes all existing records, migrationInfo in state is completely overwitten by the given migrationInfo.
-func (k Keeper) OverwriteMigrationRecords(ctx sdk.Context, migrationInfo types.MigrationRecords) {
+// deleteMigrationKeys deletes all migration records with the given prefixKey.
+func (k Keeper) deleteMigrationKeys(ctx sdk.Context, prefixKey []byte) {
 	store := ctx.KVStore(k.storeKey)
-
-	// delete all existing keys
-	prefixStore := prefix.NewStore(store, types.KeyPrefixMigrationInfoBalancerPool)
+	prefixStore := prefix.NewStore(store, prefixKey)
 	iter := prefixStore.Iterator(nil, nil)
 	defer iter.Close()
 
 	for ; iter.Valid(); iter.Next() {
 		prefixStore.Delete(iter.Key())
 	}
+}
 
-	prefixStore = prefix.NewStore(store, types.KeyPrefixMigrationInfoCLPool)
-	iter = prefixStore.Iterator(nil, nil)
-	defer iter.Close()
+// OverwriteMigrationRecordsAndRedirectDistrRecords sets the balancer to gamm pool migration info to the store and deletes all existing records
+// migrationInfo in state is completely overwitten by the given migrationInfo.
+// Additionally, the distribution record for the balancer pool is modified to redirect incentives to the new concentrated pool.
+func (k Keeper) OverwriteMigrationRecordsAndRedirectDistrRecords(ctx sdk.Context, migrationInfo types.MigrationRecords) error {
+	store := ctx.KVStore(k.storeKey)
 
-	for ; iter.Valid(); iter.Next() {
-		prefixStore.Delete(iter.Key())
+	// delete all existing keys
+	k.deleteMigrationKeys(ctx, types.KeyPrefixMigrationInfoBalancerPool)
+	k.deleteMigrationKeys(ctx, types.KeyPrefixMigrationInfoCLPool)
+
+	for _, balancerToCLPoolLink := range migrationInfo.BalancerToConcentratedPoolLinks {
+		balancerToClPoolKey := types.GetKeyPrefixMigrationInfoBalancerPool(balancerToCLPoolLink.BalancerPoolId)
+		store.Set(balancerToClPoolKey, sdk.Uint64ToBigEndian(balancerToCLPoolLink.ClPoolId))
```

Comment:
We should be using `osmoutils.MustSet` here and below to ensure there are no marshaling-related errors right?

Reply:
used osmoutils for getting and setting here https://github.com/osmosis-labs/osmosis/pull/5400/commits/46e62fc6bd262659cd7f9ab5ecc310353d68025d

note this added more complexity, and I think when values are simple uints that marshalling related errors are probably not worth the added complexity when compared to using osmoutils MustSet for custom proto structs

Reply:
I actually reverted this change since it was adding a lot of complexity for unmarshalling a uint64 value

Reply:
https://github.com/osmosis-labs/osmosis/pull/5400/commits/f9fef541ef3443b1bdfeca2462288afc1fa4ada0